File: sorting/insertion/InsertionSort.java
package sorting.insertion;

import java.util.Arrays;

public class InsertionSort {
  public static void insertionSort(int[] arr) {
    int value, hole;

    for (int i = 1; i < arr.length; i++) {
      value = arr[i];
      hole = i;

      // shift values to the right
      while (hole > 0 && arr[hole - 1] > value) {
        arr[hole] = arr[hole - 1];

        hole--;
      }

      // then current value into the hole position
      arr[hole] = value;
    }
  }

  public static void insertionSort2(int[] arr) {
    int temp, value, hole;

    for (int i = 1; i < arr.length; i++) {
      value = arr[i];
      hole = i;

      // swap values
      while (hole > 0 && arr[hole - 1] > value) {
        temp = arr[hole - 1];
        arr[hole - 1] = arr[hole];
        arr[hole] = temp;

        hole--;
      }
    }
  }

  public static void main(String[] args) {
    int[] arr = { 10, 9, 6, 5, 8, -1, -6 };

    insertionSort(arr);

    System.out.println(Arrays.toString(arr));
  }
}

File: sorting/insertion/insertion_sort.rb
def insertion_sort(arr)
  len = arr.length - 1

  for i in 1..len
    value = arr[i]
    hole = i

    # shift values to the right
    # starting from hole going backwards
    # until previous is greater or hole is 0
    while hole > 0 && arr[hole - 1] > value
      arr[hole] = arr[hole - 1]
      hole -= 1
    end

    arr[hole] = value
  end

  arr
end

p insertion_sort [10, 9, 6, 5, 8, -1, -6]
